drivers/misc/morse_code/morse_code.c

```diff
@@ -0,0 +1,406 @@
+	0x0700005d, /*     R     .-.          1011101                                    101 1101 */
+	0x05000015, /*     S     ...          10101                                        1 0101 */
+	0x03000007, /*     T     -            111                                             111 */
+	0x07000057, /*     U     ..-          1010111                                    101 0111 */
+	0x09000157, /*     V     ...-         101010111                               1 0101 0111 */
+	0x09000177, /*     W     .--          101110111                               1 0111 0111 */
+	0x0b000757, /*     X     -..-         11101010111                           111 0101 0111 */
+	0x0d001d77, /*     Y     -.--         1110101110111                      1 1101 0111 0111 */
+	0x0b000775, /*     Z     --..         11101110101                           111 0111 0101 */
+	0x0f005555, /*     [     error								  */
+	0x0f005555, /*     \     error								  */
+	0x0f005555, /*     ]     error								  */
+	0x0f005555, /*     ^     error								  */
+	0x11015dd7, /*     _     ..--.-       10101110111010111             1 0101 1101 1101 0111 */
+
+	0x03000000, /* letter space [3.3]     000                                             000 */
+	0x0f0075d7, /* /CT        -.-.-       111010111010111                  111 0101 1101 0111 */
+	0x0d00175d, /* /AR        .-.-.       1011101011101                      1 0111 0101 1101 */
+};
+
+static void morse_code_cb_handler(struct k_work *item)
+{
+	struct morse_code_data *ctx = CONTAINER_OF(item, struct morse_code_data, cb_work);
+	struct morse_code_callback_data *cb_info;
+
+	if (ctx == NULL) {
+		return;
+	}
+
+	cb_info = &ctx->cb_info;
+
+	if (cb_info->callback) {
+		(cb_info->callback)(cb_info->ctx, cb_info->status);
+	}
+}
+
+static int morse_code_peak_bit_state(struct morse_code_data *ctx)
+{
+	uint32_t bit = morse_code_symbols[ctx->code_idx] & BIT(--ctx->code_bit);
+
+	if (ctx->code_bit == 0) {
+		ctx->code_fsm = MORSE_CODE_STREAM_FSM_LOAD;
+		if (ctx->bit_fsm == MORSE_CODE_BIT_FSM_LOAD_SYMBOL) {
+			++ctx->data_idx;
+		}
+	}
+
+	return bit;
+}
+
+static int morse_code_load(struct morse_code_data *ctx)
+{
+	uint8_t symbol;
+
+	if (ctx->bit_fsm == MORSE_CODE_BIT_FSM_LOAD_SYMBOL) {
+		symbol = ctx->data[ctx->data_idx];
+		LOG_DBG("Loading %c idx: %d, size: %d", symbol, ctx->data_idx, ctx->data_size);
+
+		if (ctx->data_idx + 1 < ctx->data_size
+		&&  ctx->data[ctx->data_idx]     != 0x20
+		&&  ctx->data[ctx->data_idx + 1] != 0x20) {
+			ctx->bit_fsm = MORSE_CODE_BIT_FSM_WAIT_BLANK_PHASE;
+		}
+
+		/* Sanity Check: Out of Bounds */
+		if (symbol < 0x20 || symbol >= 0x80) {
+			LOG_ERR("Character invalid.");
+
+			return -EINVAL;
+		}
+
+		symbol -= (symbol > 0x60) ? 0x40 : 0x20;
+	} else {
+		ctx->bit_fsm = MORSE_CODE_BIT_FSM_LOAD_SYMBOL;
+		symbol = MORSE_CODE_PROSIGN_LETTER_SPACE;
+	}
+
+	ctx->code_idx = symbol;
+	ctx->code_bit = (morse_code_symbols[symbol] >> 24);
```

Comment:
If my above assumption about the encoding is correct, we could be explicit here about which bits we're taking, or at least provide a comment indicating such.

```suggestion
	ctx->code_bit = (morse_code_symbols[symbol] >> 24) & 0x1f;
```

... it may even be worth bounds checking the `code_bit` value for `<= 23`?

Reply:
done!